framework/src/main/java/org/checkerframework/framework/type/AnnotatedTypeFactory.java

```diff
@@ -3355,20 +3389,62 @@ public boolean hasQualifierParameterInHierarchy(
      *     has a qualifier parameter
      */
     public Set<AnnotationMirror> getQualifierParameterHierarchies(@Nullable Element element) {
+        if (element == null || !ElementUtils.isTypeDeclaration(element)) {
+            return AnnotationUtils.createAnnotationSet();
+        }
+
+        Set<AnnotationMirror> found =
+                getSupportedAnnotationsInElementAnnotation(element, HasQualifierParameter.class);
+        Set<AnnotationMirror> hasQualifierParameterTops = AnnotationUtils.createAnnotationSet();
+        PackageElement packageElement = ElementUtils.enclosingPackage(element);
+
+        // Traverse all packages containing this element.
+        while (packageElement != null) {
+            Set<AnnotationMirror> packageDefaultTops =
+                    getSupportedAnnotationsInElementAnnotation(
+                            packageElement, HasQualifierParameter.class);
+            hasQualifierParameterTops.addAll(packageDefaultTops);
+
+            packageElement = ElementUtils.parentPackage(packageElement, elements);
+        }
+
+        Set<AnnotationMirror> noQualifierParamClasses =
+                getSupportedAnnotationsInElementAnnotation(element, NoQualifierParameter.class);
+        for (AnnotationMirror anno : hasQualifierParameterTops) {
+            if (!AnnotationUtils.containsSame(noQualifierParamClasses, anno)) {
+                found.add(anno);
+            }
+        }
+
+        return found;
+    }
+
+    /**
+     * Returns the set of supported annotations listed inside a specified annotation written on the
```

Comment:
`listed inside` is unclear.  I think you mean this returns a set of annotation mirrors corresponding to the annotation classes listed in the value element of annotation with class elementAnnotation on element.